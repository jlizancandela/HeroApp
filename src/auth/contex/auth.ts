import { createContext } from 'react';

interface AuthContextType {
  state: { user: string | null; logedIn: boolean };
  onLogin: (user: string) => void;
  onLogout: () => void;
}

// Creamos el contexto con un valor predeterminado
export const AuthContext = createContext<AuthContextType>({
  state: { user: null, logedIn: false },
  onLogin: () => {},
  onLogout: () => {},
});
