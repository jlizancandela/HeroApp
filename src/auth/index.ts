export { LoginPage } from './pages/login';
export { AuthProvider } from './contex/AuthProvider';
export { AuthContext } from './contex/auth';
export { types } from './reducer/types';
export { AuthReducer } from './reducer/Authreducer';
