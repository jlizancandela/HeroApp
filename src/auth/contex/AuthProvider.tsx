import { useCallback, useMemo, useReducer } from 'react';
import { AuthContext } from './auth';
import { AuthReducer } from '../reducer/Authreducer';
import { types } from '../reducer/types';

interface AuthContextType {
  state: { user: string | null; logedIn: boolean }; // Cambia 'any' por el tipo correcto
  onLogin: (user: string) => void;
  onLogout: () => void;
}

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const initialState = { user: null, logedIn: false };

  const inicialicer = () => JSON.parse(localStorage.getItem('login') || '{}');

  const [state, dispatch] = useReducer(AuthReducer, initialState, inicialicer);

  const onLogin = useCallback((user: string) => {
    dispatch({ type: types.LOGIN, payload: user });
    localStorage.setItem('login', JSON.stringify({ user, logedIn: true }));
  }, []);

  const onLogout = useCallback(() => {
    dispatch({ type: types.LOGOUT, payload: null });
    localStorage.removeItem('login');
  }, []);

  const Provider = useMemo<AuthContextType>(() => ({ state, onLogin, onLogout }), [state, onLogin, onLogout]);

  return <AuthContext.Provider value={Provider}>{children}</AuthContext.Provider>;
};
