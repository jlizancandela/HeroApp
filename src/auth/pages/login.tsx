import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../contex/auth";

export const LoginPage = () => {
  const navigate = useNavigate();
  const { onLogin } = useContext(AuthContext);

  const handleLogin = () => {
    onLogin("Javier");
    navigate("/");
  };

  return (
    <section className="d-flex flex-row align-items-center justify-content-center vh-100">
      <div className="d-flex  flex-column align-items-center justify-content-center gap-3 border border-black p-5">
        <h1>LoginPage</h1>
        <button type="button" className="btn btn-primary" onClick={handleLogin}>
          Login
        </button>
      </div>
    </section>
  );
};
