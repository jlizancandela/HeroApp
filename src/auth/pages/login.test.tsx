import { LoginPage } from "./login";
import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { Mock } from "vitest";
import { AuthContext } from "../contex/auth";

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: vi.fn(),
  };
});

describe("LoginPage", () => {
  let mockedHandleLogout: Mock;
  let mockedNavigate: Mock;

  beforeEach(() => {
    mockedHandleLogout = vi.fn();
    mockedNavigate = vi.fn();
    (useNavigate as Mock).mockReturnValue(mockedNavigate);
  });

  it("should render correctly", () => {
    const { container } = render(<LoginPage />);
    expect(container).toMatchSnapshot();
  });

  it("should redirect to home", () => {
    const value = {
      state: { user: "uknown", logedIn: false },
      onLogin: mockedHandleLogout,
      onLogout: vi.fn(),
    };
    render(
      <MemoryRouter initialEntries={["/login"]}>
        <AuthContext.Provider value={value}>
          <LoginPage />
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    const button = screen.getByRole("button", { name: /login/i });
    fireEvent.click(button);
    expect(mockedHandleLogout).toHaveBeenCalled();
    expect(mockedNavigate).toHaveBeenCalledWith("/");
  });
});
