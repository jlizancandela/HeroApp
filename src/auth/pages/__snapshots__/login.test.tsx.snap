// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > should render correctly 1`] = `
<div>
  <section
    class="d-flex flex-row align-items-center justify-content-center vh-100"
  >
    <div
      class="d-flex  flex-column align-items-center justify-content-center gap-3 border border-black p-5"
    >
      <h1>
        LoginPage
      </h1>
      <button
        class="btn btn-primary"
        type="button"
      >
        Login
      </button>
    </div>
  </section>
</div>
`;
